apps/labs/posts/multi-dimensional-sparse-arrays.md

```diff
@@ -0,0 +1,225 @@
+This laid out the planned workflow, and roughly outlined the distribution of tasks across several PRs (Pull Requests) that would be submitted over time.
+
+The implementation was carried out in phases, with each phase focusing on extending a predefined set of methods to n-dimensions. These phases structured the contents of each PR.
+
+The first stage involved developing the core constructors required to initialise an n-D COO array, as well as enabling conversions between COO and dense formats. This was followed by extending basic methods such as `reshape`, `transpose`, `eliminate_zeros` etc. to n-D. Next, element-wise arithmetic operations—addition and subtraction—were implemented. At the time, this replicated the pre-existing 1D/2D sparse array implementation of element-wise operations, where we only dealt with arrays of the same shape (this behavior was later updated when we introduced broadcasting for sparse arrays, allowing element-wise operations for all broadcastable arrays to align with NumPy's handling of such methods).
+
+By the end of this stage, I had become fairly confident that the entire COO codebase could be extended to n-dimensions.
+
+I then moved on to the second stage, which involved tackling the fundamental operations in linear algebra, `matmul`, `dot`, and `tensordot`.
+
+Most functions in SciPy's sparse module fall into one of two categories: they are either purely Python-based or they involve compiled code. Compiled code is introduced when a significant speed boost is essential. However, increasing the amount of compiled code also increases the size of binaries, which can lead to portability issues and larger memory footprints. To add on, another issue with compiled code is maintenance. Not many people want to jump down into C++! Fortunately, Python's numeric library, NumPy, has been optimised so heavily that it can at times deliver as good of a performance as compiled code, if not faster, while also avoiding the drawbacks of increased memory usage associated with binaries.
+
+Given these considerations, the implementation of methods in this stage required careful thought. In the next section, I will walk through the key decisions made during the implementation of the `matmul`, `dot`, and `tensordot` methods. But first, I will introduce a concept called broadcasting.
+
+### What is broadcasting?
+
+Broadcasting is a technique used to expand arrays of different shapes to a given shape, making them compatible for element-wise operations. This process starts from the rightmost dimension and works leftwards. If both arrays have matching sizes in a particular dimension, we move on to the next. If one of the dimensions has size 1, it is "stretched" or repeated to match the other.
+
+<p align="center">
+  <img
+    alt="Visual illustration of broadcasting."
+    src="/posts/multi-dimensional-sparse-arrays/broadcasting.png"
+    width="67%"
+  />
+</p>
+<i align="center">
+  Retrieved from NumPy Documentation.
```

Comment:
Suggestion: link to https://numpy.org/doc/stable/user/basics.broadcasting.html

Reply:
Done.